rust/src/coordinator_async.rs

```diff
@@ -0,0 +1,275 @@
+use std::{
+    cmp::Ordering,
+    collections::{HashMap, VecDeque},
+    default::Default,
+    mem,
+    sync::Arc,
+};
+
+use sodiumoxide::{self};
+
+use crate::{
+    coordinator::{MaskDict, ProtocolEvent, RoundSeed},
+    crypto::ByteObject,
+    mask::{Aggregation, BoundType, DataType, GroupType, MaskConfig, ModelType},
+    message::{MessageOpen, MessageOwned, PayloadOwned},
+    message_processing::{MessageSink, MessageValidator, SumValidationData},
+    CoordinatorPublicKey,
+    CoordinatorSecretKey,
+    InitError,
+    PetError,
+    SeedDict,
+    SumDict,
+};
+use tokio::{
+    sync::{broadcast, mpsc},
+    time::Duration,
+};
+
+#[derive(Debug)]
+pub struct Start;
+#[derive(Debug)]
+pub struct Idle;
+#[derive(Debug)]
+pub struct Sum;
+#[derive(Debug)]
+pub struct Update {
+    sum_dict: Option<Arc<SumDict>>,
+}
+#[derive(Debug)]
+pub struct Sum2;
+
+// error state
+
+#[derive(Debug)]
+pub struct State<S> {
+    _inner: S,
+    // coordinator state
+    coordinator_state: CoordinatorState,
+    // message rx
+    message_rx: tokio::sync::mpsc::UnboundedReceiver<()>,
+}
+
+// Functions that are available for all states
+impl<S> State<S> {
+    fn message_opener(&self) -> MessageOpen<'_, '_> {
+        MessageOpen {
+            recipient_pk: &self.coordinator_state.pk,
+            recipient_sk: &self.coordinator_state.sk,
+        }
+    }
+
+    fn message_open(&self, message: Vec<u8>) -> Result<MessageOwned, PetError> {
+        self.message_opener()
+            .open(&message)
+            .map_err(|_| PetError::InvalidMessage)
+    }
+
+    async fn next_message(&mut self) -> Result<MessageOwned, PetError> {
+        let message = match self.message_rx.recv().await {
+            Some(message) => message,
+            None => panic!("all message senders dropped!"),
+        };
+        self.message_open(vec![1, 2, 34]) // dummy value
+    }
+}
+
+impl State<Start> {
+    pub fn new() -> Result<(tokio::sync::mpsc::UnboundedSender<()>, State<Start>), InitError> {
+        // crucial: init must be called before anything else in this module
+        sodiumoxide::init().or(Err(InitError))?;
+
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel::<()>();
+
+        Ok((
+            tx,
+            State {
+                _inner: Start,
+                coordinator_state: CoordinatorState {
+                    seed: RoundSeed::generate(),
+                    ..Default::default()
+                },
+                message_rx: rx,
+            },
+        ))
+    }
+
+    pub async fn next(self) -> State<Idle> {
```

Comment:
Could you explain where `State.next` is called and why it needs to be async? I don't see it.

Reply:
There will be a `run` that calls `next()`. Something like this:

```rust
async fn run() {
    let (_, state_start) = State::new().unwrap(); // Start
    let mut state_idle = state_start.next(); // Idle

    loop {
        let state_sum = state_idle.next().await; // Sum
        let state_update = state_sum.next().await; // Update
        let state_sum2 = state_update.next().await; // Sum2
        state_idle = state_sum2.next().await; // Idle
    }
}
```

You are right, the `next` method in the `Idle` state doesn't have to be async.